README.md
BICIB: Brainfuck Interpreter/Compiler In Brainfuck
=================================================

## About this project

BICIB aims to provide a easy-to-use tool set for brainfuck,
and it is also written in brainfuck.

BICIB provides dockerized brainfuck interpreter and compiler.
Currently only docker-capable x86_64-linux is supported.  For details
please check [docker](https://github.com/docker/docker).

## About brainfuck

Brainfuck is a esoteric programming language with only eight
instructions which can be viewed as the ability of a Turing machine. 
These instructions are(assume **head** is of type char\* in C):

<table>
<tr>
<td> symbol</td>
<td> action</td>
<td> C</td>
</tr>
<tr>
<td> "&lt;"(ASCII 0x3c)</td>
<td> move the head to the left square</td>
<td> head--</td>
</tr>
<tr>
<td> ">"(ASCII 0x3e)</td>
<td> move the head to the right square</td>
<td> head++</td>
</tr>
<tr>
<td> "+"(ASCII 0x2b)</td>
<td> increase the current square by 1</td>
<td> (*head)++</td>
</tr>
<tr>
<td> "-"(ASCII 0x2d)</td>
<td> decrease the current square by 1</td>
<td> (*head)--</td>
</tr>
<tr>
<td> ","(ASCII 0x2c)</td>
<td> write to the current square</td>
<td> *head = getchar() </td>
</tr>
<tr>
<td> "."(ASCII 0x2e)</td>
<td> read from the current square</td>
<td> putchar(*head)</td>
</tr>
<tr>
<td> "\["(ASCII 0x5b)</td>
<td> loop until the current square is 0</td>
<td> while(*head){</td>
</tr>
<tr>
<td> "\]"(ASCII 0x5d)</td>
<td> the end of the loop</td>
<td> }</td>
</tr>
</table>

It may be a little bit confusing that the "," instruction
performs a write by **getchar()** and the "." instruction 
performs a read by **putchar()**.  But notice that read/write
is the action of an operator of Turing machine, and the 
put/getchar() functions are the actions from the terminal's
point of view, which is mere a simulation to Turing machine.

More detailed history and references can easily be found at 
[Wikipedia](http://en.wikipedia.org/wiki/Brainfuck) and
[esolangs](http://esolangs.org/wiki/Brainfuck).

## Some implementation issues

Obviously, we have no way to interact with the modern computer systems
in such programming model.  Brainfuck supports neither file operations,
nor object linking.  These constraints make it hard to produce a 
self-hosted brainfuck interpreter or compiler.

We write the two programs in brainfuck.  In other words, these programs
should be viewed as Turing machines.  There is no random-accessed memory
any more, but only a tape now.

As a result, we implement a small piece of assembly code as a general 
entry point, which is responsible for loading the files specified in the
prompt into a contiguous region on the tape of the simulated machine.  
The code will then be processed by the main part of interpreter/compiler,
writen in brainfuck.

Currently(version 0.2) we have the compiler self-hosted but still use 
C interpreter to run brainfuck programs.

Getting started
===============

The only thing you have to make sure is that you have docker working 
properly.  For those who don't know how to install docker, reference
[this document](http://docs.docker.com/installation/).  Most of the 
GNU/Linux distributions and MAC are already supported, so it shouldn't
be a problem.

After you have a docker environment, you may first clone this project,
```bash
git clone git@github.com:NonerKao/BICIB.git
cd BICIB
```

And then build the container,
```bash
docker build -t bicib .
```

After the container is built, you may run it following [the guide](http://docs.docker.com/userguide/),
or briefly type the following instruction,
```bash
docker run --rm -it bicib bfi /root/BICIB/brainfuck/examples/hw.b
```

Have fun!

Licencing
=========

BICIB is licenced under [GPLv3](http://www.gnu.org/licenses/gpl-3.0.html).

Contact Me
==========

[Quey-Liang Kao](http://noner.scopelab.cs.nthu.edu.tw)


CHANGELOG.md
# Changelog

## 0.2 (2015-03-08)

#### Self-hosted Compiler

- now available
- support up to 676 loops

## 0.1 (2015-03-02)

#### Limitation
- a tape with 32768 cells

#### Implementation
- bootstrapping in C


brainfuck/interpreter.c
/* interpreter.c -- bootstrap interpreter for brainfuck
   Copyright 2015 Quey-Liang Kao.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2, or (at your option) any
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, 51 Franklin Street - Fifth Floor, Boston, MA 02110-1301, USA.*/

#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>

/*
 * "<", 0x3c, move left
 * ">", 0x3e, move right
 * ",", 0x2c, read from stdin
 * ".", 0x2e, write to tap(buffered stdout)
 * "[", 0x5b, loop until current position goes zero
 * "]", 0x5d, end loop
 * "+", 0x2b, add current position by 1
 * "-", 0x2d, substract current position by 1
 */

#define SS 1024
#define TS 32768
char tape[TS];
int head = 0;
void process(char ch);

void startwhile(char* stack, int index, char* jmp){
	char ch;
	int j = 0;
	int i;
	while(1){
		ch = stack[j];
		if(ch == '['){
			if(!tape[head]){
				j = jmp[j];
				if(j >= index-1) return;
			}
		}
		else if(ch == ']'){
			j = jmp[j];
		}
		else{
			process(ch);
		}
		j++;
	}
}

void Esys(void){
	printf("syntax error!");
	exit(0);
}

FILE* fp;

void process(char ch){
	char stack[SS];
	int index = 0;
	int diff = 0;
	char starts[SS];
	char jmps[SS];
	int top = 0;
	char temp;
	switch(ch){
		case '<':
			head = (head+TS-1)%TS;
			break;
		case '>':
			head = (head+1)%TS;
			break;
		case ',':
			tape[head] = getchar();
			break;
		case '.':
			printf("%c", tape[head]);
			break;
		case '[':
			index = 0;
			top = 0;
			starts[top++] = index;
			stack[index++] = '[';
			diff = 1;
			while(fscanf(fp, "%c", &temp) == 1){
				stack[index++] = temp;
				if(temp == '['){
					starts[top++] = index-1;
					diff++;
				}
				else if(temp == ']'){
					jmps[starts[top-1]] = index-1;
					jmps[index-1] = starts[top-1]-1;
					top--;
					diff--;
				}
				if(!diff) break;
			}
			if(temp != ']'){
				Esys();
			}
			startwhile(stack, index, jmps);
			break;
		case ']':
			Esys();
			break;
		case '+':
			tape[head]++;
			break;
		case '-':
			tape[head]--;
			break;
		default:
			break;
	}
}

int main(int argc, char* argv[]){
	char ch;
	if(argc != 2){
		printf("usage: %s <source>\n", argv[0]);
		return 0;
	}
	fp = fopen(argv[1], "r");
	while(fscanf(fp, "%c", &ch) == 1){
		process(ch);
	}
	fflush(0);
	fclose(fp);
	return 0;
}
